File: bindings/go/stores/mysql/mysql.go
package mysql

import (
	"bytes"
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"sync/atomic"

	prolly "build.crab/prolly-go"
	_ "github.com/go-sql-driver/mysql"
	mysqldriver "github.com/go-sql-driver/mysql"
)

type Options struct {
	AdapterName     string
	ReadParallelism uint32
}

type Store struct {
	db      *sql.DB
	options Options
	owned   bool
	closed  atomic.Bool
}

func New(db *sql.DB, options Options) *Store {
	if strings.TrimSpace(options.AdapterName) == "" {
		options.AdapterName = "mysql-v1"
	}
	if options.ReadParallelism == 0 {
		options.ReadParallelism = 16
	}
	return &Store{db: db, options: options}
}

func Open(dataSourceName string, options Options) (*Store, error) {
	db, err := sql.Open("mysql", dataSourceName)
	if err != nil {
		return nil, mysqlError("open", err)
	}
	store := New(db, options)
	store.owned = true
	return store, nil
}

func (s *Store) InitializeSchema(ctx context.Context) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	for _, statement := range strings.Split(Schema, ";") {
		if strings.TrimSpace(statement) == "" {
			continue
		}
		if _, err := s.db.ExecContext(ctx, statement); err != nil {
			return mysqlError("initialize_schema", err)
		}
	}
	return nil
}

func (s *Store) Close() error {
	if s == nil || s.closed.Swap(true) || !s.owned || s.db == nil {
		return nil
	}
	return s.db.Close()
}

func (s *Store) Descriptor(ctx context.Context) (prolly.StoreDescriptor, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreDescriptor{}, err
	}
	return prolly.StoreDescriptor{
		ProtocolMajor: prolly.StoreProtocolMajor, AdapterName: s.options.AdapterName, Provider: "mysql", SchemaVersion: 1,
		Capabilities: prolly.StoreCapabilities{
			NativeBatchReads: true, AtomicBatchWrites: true, NodeScan: true, Hints: true,
			AtomicNodesAndHint: true, RootScan: true, RootCompareAndSwap: true,
			Transactions: true, ReadParallelism: s.options.ReadParallelism,
		},
	}, nil
}

func (s *Store) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.db, selectNode, key)
}
func (s *Store) PutNode(ctx context.Context, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.db.ExecContext(ctx, upsertNode, key, value)
	return mysqlError("put_node", err)
}
func (s *Store) DeleteNode(ctx context.Context, key []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.db.ExecContext(ctx, deleteNode, key)
	return mysqlError("delete_node", err)
}
func (s *Store) BatchNodes(ctx context.Context, mutations []prolly.NodeMutation) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	return s.withTx(ctx, "batch_nodes", func(tx *sql.Tx) error { return applyNodes(ctx, tx, mutations) })
}
func (s *Store) PublishNodes(ctx context.Context, publication prolly.NodePublication) error {
	return prolly.PublishNodesWithGeneralPath(ctx, s, publication)
}

func (s *Store) BatchGetNodesOrdered(ctx context.Context, keys [][]byte) ([]prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	result := make([]prolly.OptionalBytes, len(keys))
	for index, key := range keys {
		value, err := queryOptional(ctx, s.db, selectNode, key)
		if err != nil {
			return nil, err
		}
		result[index] = value
	}
	return result, nil
}
func (s *Store) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	rows, err := s.db.QueryContext(ctx, `SELECT cid FROM prolly_nodes ORDER BY cid`)
	if err != nil {
		return nil, mysqlError("list_nodes", err)
	}
	defer rows.Close()
	var result [][]byte
	for rows.Next() {
		var value []byte
		if err := rows.Scan(&value); err != nil {
			return nil, mysqlError("list_nodes_scan", err)
		}
		result = append(result, clone(value))
	}
	return result, mysqlError("list_nodes_rows", rows.Err())
}

func (s *Store) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.db, selectHint, namespace, key)
}
func (s *Store) PutHint(ctx context.Context, namespace, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.db.ExecContext(ctx, upsertHint, namespace, key, value)
	return mysqlError("put_hint", err)
}
func (s *Store) BatchPutNodesWithHint(ctx context.Context, nodes []prolly.NodeEntry, namespace, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	return s.withTx(ctx, "batch_nodes_hint", func(tx *sql.Tx) error {
		for _, node := range nodes {
			if _, err := tx.ExecContext(ctx, upsertNode, node.Key, node.Value); err != nil {
				return mysqlError("batch_node", err)
			}
		}
		_, err := tx.ExecContext(ctx, upsertHint, namespace, key, value)
		return mysqlError("batch_hint", err)
	})
}

func (s *Store) GetRootManifest(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.db, selectRoot, name)
}
func (s *Store) PutRootManifest(ctx context.Context, name, manifest []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.db.ExecContext(ctx, upsertRoot, name, manifest)
	return mysqlError("put_root", err)
}
func (s *Store) DeleteRootManifest(ctx context.Context, name []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.db.ExecContext(ctx, deleteRoot, name)
	return mysqlError("delete_root", err)
}
func (s *Store) CompareAndSwapRootManifest(ctx context.Context, name []byte, expected, replacement prolly.OptionalBytes) (prolly.RootCASResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.RootCASResult{}, err
	}
	var result prolly.RootCASResult
	err := s.withTx(ctx, "root_cas", func(tx *sql.Tx) error {
		current, err := queryOptional(ctx, tx, selectRootForUpdate, name)
		if err != nil {
			return err
		}
		if !optionalEqual(current, expected) {
			result.Current = current
			return nil
		}
		if err := writeRoot(ctx, tx, name, replacement); err != nil {
			return err
		}
		result = prolly.RootCASResult{Applied: true, Current: replacement.Clone()}
		return nil
	})
	return result, err
}
func (s *Store) ListRootManifests(ctx context.Context) ([]prolly.NamedStoreRoot, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	rows, err := s.db.QueryContext(ctx, `SELECT name, manifest FROM prolly_roots ORDER BY name`)
	if err != nil {
		return nil, mysqlError("list_roots", err)
	}
	defer rows.Close()
	var result []prolly.NamedStoreRoot
	for rows.Next() {
		var name, manifest []byte
		if err := rows.Scan(&name, &manifest); err != nil {
			return nil, mysqlError("list_roots_scan", err)
		}
		result = append(result, prolly.NamedStoreRoot{Name: clone(name), Manifest: clone(manifest)})
	}
	return result, mysqlError("list_roots_rows", rows.Err())
}

func (s *Store) CommitTransaction(ctx context.Context, nodes []prolly.NodeMutation, conditions []prolly.RootCondition, roots []prolly.RootWrite) (prolly.StoreTransactionResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreTransactionResult{}, err
	}
	var result prolly.StoreTransactionResult
	err := s.withTx(ctx, "transaction", func(tx *sql.Tx) error {
		for _, condition := range conditions {
			current, err := queryOptional(ctx, tx, selectRootForUpdate, condition.Name)
			if err != nil {
				return err
			}
			if !optionalEqual(current, condition.Expected) {
				result.Conflict = &prolly.StoreTransactionConflict{Name: clone(condition.Name), Expected: condition.Expected.Clone(), Current: current}
				return nil
			}
		}
		if err := applyNodes(ctx, tx, nodes); err != nil {
			return err
		}
		for _, root := range roots {
			if err := writeRoot(ctx, tx, root.Name, root.Replacement); err != nil {
				return err
			}
		}
		result.Applied = true
		return nil
	})
	return result, err
}

type queryer interface {
	QueryRowContext(context.Context, string, ...any) *sql.Row
}
type execer interface {
	ExecContext(context.Context, string, ...any) (sql.Result, error)
}

func queryOptional(ctx context.Context, target queryer, query string, args ...any) (prolly.OptionalBytes, error) {
	var value []byte
	err := target.QueryRowContext(ctx, query, args...).Scan(&value)
	if errors.Is(err, sql.ErrNoRows) {
		return prolly.MissingBytes(), nil
	}
	if err != nil {
		return prolly.OptionalBytes{}, mysqlError("query", err)
	}
	return prolly.PresentBytes(value), nil
}
func applyNodes(ctx context.Context, tx *sql.Tx, mutations []prolly.NodeMutation) error {
	for _, mutation := range mutations {
		var err error
		if mutation.Value.Present {
			_, err = tx.ExecContext(ctx, upsertNode, mutation.Key, mutation.Value.Value)
		} else {
			_, err = tx.ExecContext(ctx, deleteNode, mutation.Key)
		}
		if err != nil {
			return mysqlError("node_mutation", err)
		}
	}
	return nil
}
func writeRoot(ctx context.Context, target execer, name []byte, value prolly.OptionalBytes) error {
	var err error
	if value.Present {
		_, err = target.ExecContext(ctx, upsertRoot, name, value.Value)
	} else {
		_, err = target.ExecContext(ctx, deleteRoot, name)
	}
	return mysqlError("write_root", err)
}
func (s *Store) withTx(ctx context.Context, operation string, call func(*sql.Tx) error) error {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return mysqlError(operation+"_begin", err)
	}
	defer tx.Rollback()
	if err := call(tx); err != nil {
		return err
	}
	return mysqlError(operation+"_commit", tx.Commit())
}
func optionalEqual(left, right prolly.OptionalBytes) bool {
	return left.Present == right.Present && (!left.Present || bytes.Equal(left.Value, right.Value))
}
func clone(value []byte) []byte { return append([]byte(nil), value...) }
func (s *Store) ready(ctx context.Context) error {
	if ctx != nil && ctx.Err() != nil {
		return ctx.Err()
	}
	if s == nil || s.db == nil {
		return &prolly.StoreError{Code: "invalid_store", Message: "MySQL database is nil"}
	}
	if s.closed.Load() {
		return &prolly.StoreError{Code: "closed", Message: "MySQL store is closed"}
	}
	return nil
}
func mysqlError(operation string, err error) error {
	if err == nil {
		return nil
	}
	retryable, providerCode := false, ""
	var driverErr *mysqldriver.MySQLError
	if errors.As(err, &driverErr) {
		providerCode = fmt.Sprint(driverErr.Number)
		switch driverErr.Number {
		case 1040, 1205, 1213, 2006, 2013:
			retryable = true
		}
	}
	return &prolly.StoreError{Code: "mysql", Message: fmt.Sprintf("%s: %v", operation, err), Retryable: retryable, ProviderCode: providerCode, Cause: err}
}

var _ prolly.RemoteStore = (*Store)(nil)

File: bindings/ruby/stores/mysql/lib/prolly/store/mysql.rb
# frozen_string_literal: true

require 'mysql2'
require 'prolly'

module Prolly
  class MysqlRemoteStore < ForeignRemoteStore
    CREATE_SCHEMA = [
      'CREATE TABLE IF NOT EXISTS prolly_nodes (cid VARBINARY(32) PRIMARY KEY, node LONGBLOB NOT NULL)',
      'CREATE TABLE IF NOT EXISTS prolly_hints (namespace VARBINARY(255) NOT NULL, `key` VARBINARY(255) NOT NULL, value LONGBLOB NOT NULL, PRIMARY KEY(namespace, `key`))',
      'CREATE TABLE IF NOT EXISTS prolly_roots (name VARBINARY(255) PRIMARY KEY, manifest LONGBLOB NOT NULL)'
    ].freeze
    UPSERT_NODE = 'INSERT INTO prolly_nodes VALUES (?, ?) AS new ON DUPLICATE KEY UPDATE node=new.node'
    UPSERT_HINT = 'INSERT INTO prolly_hints VALUES (?, ?, ?) AS new ON DUPLICATE KEY UPDATE value=new.value'
    UPSERT_ROOT = 'INSERT INTO prolly_roots VALUES (?, ?) AS new ON DUPLICATE KEY UPDATE manifest=new.manifest'

    def initialize(client)
      raise ArgumentError, 'client must be Mysql2::Client' unless client.is_a?(Mysql2::Client)

      @client = client
      @mutex = Mutex.new
      @closed = false
    end

    def initialize_schema = synchronize { CREATE_SCHEMA.each { |sql| @client.query(sql) } }
    def close = @closed = true

    def descriptor
      capabilities = StoreCapabilitiesRecord.new(
        native_batch_reads: false, atomic_batch_writes: true, node_scan: true,
        hints: true, atomic_nodes_and_hint: true, root_scan: true,
        root_compare_and_swap: true, transactions: true, read_parallelism: 1
      )
      limits = StoreLimitsRecord.new(
        max_batch_read_items: nil, max_batch_write_items: nil,
        max_transaction_operations: nil, max_node_bytes: nil
      )
      value = StoreDescriptorRecord.new(
        protocol_major: STORE_PROTOCOL_MAJOR, adapter_name: 'mysql-v1', provider: 'mysql',
        schema_version: 1, capabilities: capabilities, limits: limits
      )
      StoreDescriptorResultRecord.new(value: value, error: nil)
    end

    def get_node(cid)
      optional_result(query('SELECT node FROM prolly_nodes WHERE cid=?', key(cid, 32, 'node CID')))
    rescue StandardError => error
      OptionalBytesResultRecord.new(value: optional(nil), error: store_error(error))
    end
    def put_node(cid, value)
      write(UPSERT_NODE, key(cid, 32, 'node CID'), value)
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def delete_node(cid)
      write('DELETE FROM prolly_nodes WHERE cid=?', key(cid, 32, 'node CID'))
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def batch_nodes(operations)
      transaction do
        operations.each do |item|
          cid = key(item.key, 32, 'node CID')
          item.value.present ? execute(UPSERT_NODE, cid, item.value.value) : execute('DELETE FROM prolly_nodes WHERE cid=?', cid)
        end
      end
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def batch_get_nodes_ordered(cids)
      values = synchronize { cids.map { |cid| optional(query_unlocked('SELECT node FROM prolly_nodes WHERE cid=?', key(cid, 32, 'node CID'))) } }
      OptionalBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      OptionalBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def list_node_cids
      values = synchronize { rows('SELECT cid FROM prolly_nodes ORDER BY cid').map { |row| row[0].b } }
      BytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      BytesListResultRecord.new(values: [], error: store_error(error))
    end

    def get_hint(namespace, hint_key)
      optional_result(query('SELECT value FROM prolly_hints WHERE namespace=? AND `key`=?', key(namespace, 255, 'hint namespace'), key(hint_key, 255, 'hint key')))
    rescue StandardError => error
      OptionalBytesResultRecord.new(value: optional(nil), error: store_error(error))
    end

    def put_hint(namespace, hint_key, value)
      write(UPSERT_HINT, key(namespace, 255, 'hint namespace'), key(hint_key, 255, 'hint key'), value)
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def batch_put_nodes_with_hint(nodes, namespace, hint_key, value)
      transaction do
        nodes.each { |node| execute(UPSERT_NODE, key(node.key, 32, 'node CID'), node.value) }
        execute(UPSERT_HINT, key(namespace, 255, 'hint namespace'), key(hint_key, 255, 'hint key'), value)
      end
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def get_root_manifest(name)
      optional_result(query('SELECT manifest FROM prolly_roots WHERE name=?', key(name, 255, 'root name')))
    rescue StandardError => error
      OptionalBytesResultRecord.new(value: optional(nil), error: store_error(error))
    end

    def put_root_manifest(name, manifest)
      write(UPSERT_ROOT, key(name, 255, 'root name'), manifest)
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def delete_root_manifest(name)
      write('DELETE FROM prolly_roots WHERE name=?', key(name, 255, 'root name'))
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def compare_and_swap_root_manifest(name, expected, replacement)
      name = key(name, 255, 'root name')
      synchronize do
        transaction_unlocked([name]) do
          current = query_unlocked('SELECT manifest FROM prolly_roots WHERE name=? FOR UPDATE', name)
          unless matches?(current, expected)
            next RootCasResultRecord.new(applied: false, current: optional(current), error: nil)
          end
          write_root(name, replacement)
          RootCasResultRecord.new(applied: true, current: replacement, error: nil)
        end
      end
    rescue StandardError => error
      RootCasResultRecord.new(applied: false, current: optional(nil), error: store_error(error))
    end

    def list_root_manifests
      values = synchronize do
        rows('SELECT name,manifest FROM prolly_roots ORDER BY name').map { |row| NamedBytesRecord.new(name: row[0].b, value: row[1].b) }
      end
      NamedBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      NamedBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def commit_transaction(nodes, conditions, roots)
      names = conditions.map { |condition| key(condition.name, 255, 'root name') }.uniq.sort
      synchronize do
        transaction_unlocked(names) do
          conflict = nil
          conditions.each do |condition|
            current = query_unlocked('SELECT manifest FROM prolly_roots WHERE name=? FOR UPDATE', condition.name)
            next if matches?(current, condition.expected)

            conflict = StoreTransactionConflictRecord.new(name: condition.name, expected: condition.expected, current: optional(current))
            break
          end
          next TransactionResultRecord.new(applied: false, conflict: conflict, error: nil) if conflict

          nodes.each do |item|
            cid = key(item.key, 32, 'node CID')
            item.value.present ? execute(UPSERT_NODE, cid, item.value.value) : execute('DELETE FROM prolly_nodes WHERE cid=?', cid)
          end
          roots.each { |root| write_root(key(root.name, 255, 'root name'), root.replacement) }
          TransactionResultRecord.new(applied: true, conflict: nil, error: nil)
        end
      end
    rescue StandardError => error
      TransactionResultRecord.new(applied: false, conflict: nil, error: store_error(error))
    end

    private

    def synchronize(&block)
      raise 'MySQL store is closed' if @closed
      @mutex.synchronize(&block)
    end

    def key(value, maximum, label)
      value = value.b
      raise ArgumentError, "#{label} exceeds #{maximum} bytes" if value.bytesize > maximum
      value
    end

    def execute(sql, *values)
      statement = @client.prepare(sql)
      statement.execute(*values.map(&:b), as: :array).to_a
    ensure
      statement&.close
    end

    def rows(sql) = @client.query(sql, as: :array, cast: false).to_a
    def query(sql, *values) = synchronize { query_unlocked(sql, *values) }
    def query_unlocked(sql, *values) = execute(sql, *values).first&.first&.b

    def write(sql, *values)
      synchronize { execute(sql, *values) }
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def transaction(&block) = synchronize { transaction_unlocked([], &block) }

    def transaction_unlocked(lock_names)
      acquired = []
      lock_names.each do |name|
        raise 'MySQL root lock timed out' unless execute("SELECT GET_LOCK(CONCAT('prolly:', HEX(?)), 10)", name).first.first == 1
        acquired << name
      end
      @client.query('BEGIN')
      begin
        value = yield
        @client.query('COMMIT')
        value
      rescue Exception
        @client.query('ROLLBACK') rescue nil
        raise
      ensure
        acquired.reverse_each { |name| execute("SELECT RELEASE_LOCK(CONCAT('prolly:', HEX(?)))", name) }
      end
    end

    def optional(value) = OptionalBytesRecord.new(present: !value.nil?, value: value || ''.b)
    def optional_result(value) = OptionalBytesResultRecord.new(value: optional(value), error: nil)
    def unit = UnitResultRecord.new(error: nil)
    def matches?(current, expected) = expected.present ? current == expected.value : current.nil?
    def write_root(name, replacement) = replacement.present ? execute(UPSERT_ROOT, name, replacement.value) : execute('DELETE FROM prolly_roots WHERE name=?', name)

    def store_error(error)
      StoreErrorRecord.new(
        code: error.is_a?(ArgumentError) ? 'invalid_argument' : 'internal',
        message: 'MySQL provider operation failed', retryable: false,
        provider_code: error.respond_to?(:error_number) ? error.error_number.to_s : nil
      )
    end
  end
end
